### PR TITLE
Q073/Q074

### LINKED ISSUE #240: Validation (2017): Q073 (Type of Purchaser)
**Edit #:** Q073
**Type:** Macro
**Category:** The Macro Quality Edit Report contains the following edits and is generated by the FFIEC
**Column:** Type of Purchaser

If 250 or greater loans are reported with purpose of loan = 1, action taken type = 1 or 6, property type = 1 or 2, and loan type = 2 or 3, then of these loans, the percentage of loans sold should be > 20%.

Additional resources:
- [FFIEC's 2016 HMDA File Spec](https://www.ffiec.gov/hmda/pdf/spec2016.pdf)
- [FFIEC's 2016 HMDA Edits](https://www.ffiec.gov/hmda/pdf/edit2016.pdf)
- [CFPB's 2017 HMDA File Spec](http://www.consumerfinance.gov/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf)

### LINKED ISSUE #241: Validation (2017): Q074 (Type of Purchaser)
**Edit #:** Q074
**Type:** Macro
**Category:** The Macro Quality Edit Report contains the following edits and is generated by the FFIEC
**Column:** Type of Purchaser

If 250 or greater loans are reported with purpose of loan = 3, action taken type = 1 or 6, property type = 1 or 2, and loan type = 2 or 3, then of these loans, the percentage of loans sold should be > 20%.

Additional resources:
- [FFIEC's 2016 HMDA File Spec](https://www.ffiec.gov/hmda/pdf/spec2016.pdf)
- [FFIEC's 2016 HMDA Edits](https://www.ffiec.gov/hmda/pdf/edit2016.pdf)
- [CFPB's 2017 HMDA File Spec](http://www.consumerfinance.gov/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf)

### PR DESCRIPTION
**Pull request by nickgrippin**
Closes #240 
Closes #241

### CODE REVIEW

File: validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    val relevantLoan = lar.loan.copy(propertyType = 2, loanType = 2, purpose = 1)
+    lar.copy(actionTakenType = 6, loan = relevantLoan)
+  }
+
+  val irrelevantAmount: Gen[Int] = Gen.chooseNum(1, numOfLoanApplications - 1)
+  val relevantAmount: Gen[Int] = Gen.chooseNum(numOfLoanApplications, 10000)
+
+  property(s"be valid if fewer than $numOfLoanApplications specified loans") {
+    forAll(irrelevantAmount) { (x) =>
+      val lars = larNGen(x).sample.getOrElse(Nil)
```
Comment (schbetsy) [inline validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala]: Could these just be `get`, rather than `getOrElse`?

File: validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+import hmda.model.fi.lar.LoanApplicationRegister
+import hmda.validation.rules.AggregateEditCheck
+import hmda.validation.rules.lar.`macro`.MacroEditTypes.LoanApplicationRegisterSource
+import org.scalacheck.Gen
+
+class Q073Spec extends MacroSpec {
+
+  val numOfLoanApplications = config.getInt("hmda.validation.macro.Q073.numOfLoanApplications")
+  val denialMultiplier = config.getDouble("hmda.validation.macro.Q073.numOfLarsMultiplier")
```
Comment (schbetsy) [inline validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala]: I'd name this variable `ratioOfLoansSold` or something similar.

File: validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  def relevantLar(lar: LoanApplicationRegister) = {
+    val relevantLoan = lar.loan.copy(propertyType = 2, loanType = 2, purpose = 1)
+    lar.copy(actionTakenType = 6, loan = relevantLoan)
+  }
+
+  val irrelevantAmount: Gen[Int] = Gen.chooseNum(1, numOfLoanApplications - 1)
+  val relevantAmount: Gen[Int] = Gen.chooseNum(numOfLoanApplications, 10000)
+
+  property(s"be valid if fewer than $numOfLoanApplications specified loans") {
+    forAll(irrelevantAmount) { (x) =>
```
Comment (schbetsy) [inline validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala]: Most other macro specs don't use `forAll` to run with randomly generated numbers of loans, they all just choose a few values that test the edge cases of the specific edit. 

I like that these specs cover more of the possible cases. I just want to make sure that it's okay that they don't follow the same pattern as the others, and that they'll be slower, since they're testing hundreds of cases instead of a few.

(for the record: I don't think you should have to change them to single tests; I just wanted to put that decision in the open)

File: validation/src/main/scala/hmda/validation/rules/lar/macro/Q074.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      count(lars.filter(lar =>
+        lar.loan.purpose == 3
+          && Seq(1, 6).contains(lar.actionTakenType)
+          && Seq(1, 2).contains(lar.loan.propertyType)
+          && Seq(2, 3).contains(lar.loan.loanType)))
+
+    val sold =
+      count(lars.filter(lar =>
+        lar.loan.purpose == 3
+          && lar.actionTakenType == 6
```
Comment (schbetsy) [inline validation/src/main/scala/hmda/validation/rules/lar/macro/Q074.scala]: See the comment [here](https://github.com/cfpb/hmda-platform/issues/240#issuecomment-263991392) for clarification on the logic of the edit. The count of loans sold will have to filter on the `purchaserType` field.

Comment (nickgrippin) [inline validation/src/main/scala/hmda/validation/rules/lar/macro/Q074.scala]: My bad... I had asked someone about this, but I think there might have been a mixup.

File: validation/src/main/scala/hmda/validation/rules/lar/macro/Q073.scala [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    val purchaserType =
+      count(lars.filter(lar =>
+        lar.loan.purpose == 1
+          && Seq(1, 6).contains(lar.actionTakenType)
+          && Seq(1, 2).contains(lar.loan.propertyType)
+          && Seq(2, 3).contains(lar.loan.loanType)))
+
+    val sold =
+      count(lars.filter(lar =>
+        lar.loan.purpose == 1
```
Comment (schbetsy) [inline validation/src/main/scala/hmda/validation/rules/lar/macro/Q073.scala]: There's repeated logic here for filtering these two sets. Could we just filter `lars` once, then use that to calculate the values?

File: validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  val overThreshold = numOfLoanApplications + 1
+
+  property(s"be valid if fewer than $numOfLoanApplications specified loans") {
+    val lars = larNGen(belowThreshold).sample.get
+    val validLarSource = newLarSource(lars, belowThreshold, relevantLar, irrelevantLar)
+    validLarSource.mustPass
+  }
+
+  property(s"be valid if more than $numOfLoanApplications specified loans and count > $multiplier * sold") {
+    val numOfRelevantLars = (overThreshold * multiplier).toInt + 1
```
Comment (schbetsy) [inline validation/src/test/scala/hmda/validation/rules/lar/macro/Q073Spec.scala]: We don't have to use `overThreshold`, since the rule says "250 or greater". I think just using `threshold` would be a better test that we've covered the edge case.

**Comment by schbetsy** [comment]
It's looking good. Just a few details to clean up.

File: validation/src/main/scala/hmda/validation/rules/lar/macro/Q074.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      count(lars.filter(lar =>
+        lar.loan.purpose == 3
+          && Seq(1, 6).contains(lar.actionTakenType)
+          && Seq(1, 2).contains(lar.loan.propertyType)
+          && Seq(2, 3).contains(lar.loan.loanType)))
+
+    val sold =
+      count(lars.filter(lar =>
+        lar.loan.purpose == 3
+          && lar.purchaserType != 0
```
Comment (kgudel) [inline validation/src/main/scala/hmda/validation/rules/lar/macro/Q074.scala]: I believe that the same method of removing repeated logic can be used here as in Q074
